KAFKA-19683: More test replacements [5/N]

shashankhs11 opened the pull request:
Replaced more tests in `TaskManagerTest.java`

Reviewers: Lucas Brutschy <lbrutschy@confluent.io>

shashankhs11 committed:
wip-1

shashankhs11 commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java:
```diff
@@ -3041,218 +3040,272 @@ public void shouldCommitAllActiveTasksThatNeedCommittingOnHandleRevocationWithEo
     }
 
     @Test
     public void shouldNotCommitIfNoRevokedTasksNeedCommittingWithEOSv2() {
```
This test tests the same behaviour as the above test `shouldNotCommitIfNoRevokedTasksNeedCommitting`. 
I wasn't able to understand/figure out on how the behavior varies through the different processing modes.

Just for my understanding, it would differ only if the revoked tasks needs committing right?

lucasbru replied:
yes. We could consider parametrizing on ProcessingMode here.

shashankhs11 replied:
done in ff6b3b0

shashankhs11 commented:
@lucasbru, tagging for review :)

lucasbru commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java:
```diff
@@ -2872,88 +2872,87 @@ public void shouldAddNewActiveTasks() {
-        );
-        assertThat(taskManager.standbyTaskMap(), Matchers.anEmptyMap());
-        verify(changeLogReader).enforceRestoreActive();
-        verifyNoInteractions(consumer);
```
why did you remove `verifyNoInteractions(consumer)`?

shashankhs11 replied:
added back in 7d74633

lucasbru commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java:
```diff
@@ -2872,88 +2872,87 @@ public void shouldAddNewActiveTasks() {
 
     @Test
     public void shouldNotCompleteRestorationIfTasksCannotInitialize() {
```
Wait. We should actually also handle TimeoutException in the production code. I suppose you removed TimeoutException because state updater doesn't handle it.

We should fix that, because `initializeIfNeeded` may throw a timeout exception.

If we get a TimeoutException inside addTasksToStateUpdater, we should 
```
            task.maybeInitTaskTimeoutOrThrow(now, timeoutException);
            tasks.addPendingTasksToInit(Collections.singleton(task));
            updateOrCreateBackoffRecord(task.id(), nowMs);
```
and otherwise, we should 
```
                task.clearTaskTimeout();
```
after the initialization.

And then we should bring back this behavior in the test

shashankhs11 replied:
Opened [KAFKA-19864](https://issues.apache.org/jira/browse/KAFKA-19864) to track and isolate this change -- #20829

lucasbru reviewed (commented):
thanks a lot, but it seems there is a little gap in the state updater code here, since we do not handle timeout exceptions during initializations as before. I wonder if we could fix it in this PR, as it shouldn't be a huge change?

shashankhs11 committed:
parameterize tests

shashankhs11 committed:
added missing verification

shashankhs11 commented:
@lucasbru 

> I wonder if we could fix it in this PR, as it shouldn't be a huge change?

I created [KAFKA-19864](https://issues.apache.org/jira/browse/KAFKA-19864) to track and isolate this change -- #20829 is ready for review.

The thinking behind this was that since the commits are squashed and merged to `trunk` and this PR also updates few other tests, I thought it would be better if we file isolate this change in case we need to revert for any reason in the future.

lucasbru commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java:
```diff
@@ -2872,88 +2874,88 @@ public void shouldAddNewActiveTasks() {
+        when(task00.commitNeeded()).thenReturn(true);
+        when(task00.prepareCommit(true)).thenReturn(offsets);
+        doNothing().when(task00).postCommit(true);
+        doNothing().when(task00).suspend();
```
I think void methods do no need stubbing like his

shashankhs11 replied:
oh wow! yess, nice! Remvoed in f634443

lucasbru commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java:
```diff
@@ -2872,88 +2874,88 @@ public void shouldAddNewActiveTasks() {
+        verify(task00).completeRestoration(any());
+        verify(stateUpdater).add(task00);
+        verify(tasks, never()).addTask(task00);
+        verify(consumer, never()).resume(task00.inputPartitions());
```
I guess we can remove this if we verify no interactions in the line below

shashankhs11 replied:
Done! Removed in f634443

lucasbru reviewed (commented):
Two nits, otherwise LGTM, so approving right away. Good idea to move the functional change to a separate PR

shashankhs11 committed:
minor nits

shashankhs11 committed:
revert shouldNotCompleteRestorationIfTasksCannotInitialize to original

shashankhs11 commented:
@lucasbru, PTAL 😬 
I reverted the test `shouldNotCompleteRestorationIfTasksCannotInitialize` 
back to original as we moved this to #20829 

I think we should not have any merge conflicts now.

shashankhs11 commented:
CI check failed due to a failing test. Rebased with latest trunk.
